crates/spfs/src/tracking/diff.rs

```diff
@@ -117,60 +117,70 @@ pub fn compute_diff<U1: Clone>(a: &Manifest<U1>, b: &Manifest<U1>) -> Vec<Diff<U
 
     let mut visited = HashSet::new();
     for entry in all_entries.iter() {
-        if visited.contains(&entry.path) {
+        if !visited.insert(&entry.path) {
             continue;
-        } else {
-            visited.insert(&entry.path);
-            match diff_path(a, b, &entry.path) {
-                Some(d) => changes.push(d),
-                None => tracing::debug!(
-                    // XXX this is not the only reason `diff_path` may return
-                    // None.
-                    "path was missing from both manifests during diff, this should be impossible"
-                ),
-            }
+        }
+        match diff_path(a, b, &entry.path) {
+            DiffPathResult::Diff(d) => changes.push(d),
+            DiffPathResult::PathMissingFromBothManifests => tracing::debug!(path = ?entry.path,
+                "path was missing from both manifests during diff, this should be impossible"
+            ),
+            DiffPathResult::UselessMask => tracing::debug!(path = ?entry.path,
+                "path was masked in the right manifest but didn't exist in the left"
+            ),
         }
     }
 
     changes
 }
 
+// Allow: most instances will be of the large variant; boxing is
+// counter-productive.
+#[allow(clippy::large_enum_variant)]
+enum DiffPathResult<U1, U2> {
+    Diff(Diff<U1, U2>),
+    PathMissingFromBothManifests,
+    UselessMask,
```

Comment:
I'm not entirely clear on this variant based on the naming, would you mind adding some documentation about when this result appears?

Reply:
Did you see "path was masked in the right manifest but didn't exist in the left" above?

That's what I'd add as the description.

Reply:
Oh, yes I see that does help as context